doc/man/man3/Tuner.3
.TH "Tuner" 3 "Thu Apr 16 2020" "Version v1.0" "RDA5807M Tuner API" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Tuner
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBTuner\fP (uint8_t address, uint8_t SDA, uint8_t SCL, int mcuClock)"
.br
.RI "Construct a new \fBTuner\fP object\&. "
.ti -1c
.RI "void \fBinit\fP (void)"
.br
.RI "Resets chip, enables chip, and updates both buffers\&. "
.ti -1c
.RI "void \fBupdateDevice\fP (void)"
.br
.RI "Sends updated buffers to tuner chip\&. "
.ti -1c
.RI "void \fBdumpBuffers\fP (uint8_t read[12], uint8_t write[12]) const"
.br
.RI "dumps buffers from MCU into arrays sent to function "
.ti -1c
.RI "void \fBsetStation\fP (int station)"
.br
.RI "Set the station to tune to\&. "
.ti -1c
.RI "int \fBgetStation\fP ()"
.br
.RI "Get channel chip is tuned to currently\&. "
.ti -1c
.RI "int \fBsetVolume\fP (uint8_t volume)"
.br
.RI "Set volume of chip audio amplifier\&. "
.ti -1c
.RI "uint8_t \fBgetVolume\fP ()"
.br
.RI "Get the Volume value from buffer\&. "
.ti -1c
.RI "uint8_t \fBgetSignalStrength\fP ()"
.br
.RI "Get the Signal Strength as a 7-bit int\&. "
.ti -1c
.RI "void \fBresetChip\fP (bool resetReg)"
.br
.RI "Calling this function preforms soft reset\&. "
.ti -1c
.RI "void \fBenableChip\fP ()"
.br
.RI "Enables receiver audio amplifier of chip\&. "
.ti -1c
.RI "void \fBdisableChip\fP ()"
.br
.RI "Disables receiver audio amplifier of chip\&. "
.ti -1c
.RI "void \fBsetSeekSensitiviy\fP (uint8_t threshold)"
.br
.RI "Set the Seek Sensitivity\&. "
.ti -1c
.RI "void \fBseekUp\fP ()"
.br
.RI "Command chip to preform Auto Seek Upwards\&. "
.ti -1c
.RI "void \fBseekDown\fP ()"
.br
.RI "Command chip to preform Auto Seek Downwards\&. "
.ti -1c
.RI "bool \fBseekStatus\fP (void)"
.br
.RI "Status of most recent seek\&. "
.ti -1c
.RI "bool \fBseekComplete\fP (void)"
.br
.RI "Status of if scan is complete\&. "
.ti -1c
.RI "bool \fBSTstatus\fP ()"
.br
.RI "Is the chip receiving a stereo signal? "
.ti -1c
.RI "bool \fBTunedStatus\fP ()"
.br
.RI "Is current frequency a station? "
.ti -1c
.RI "void \fBforceMono\fP (bool mono)"
.br
.RI "Force mono signal only\&. "
.ti -1c
.RI "void \fBbassBoost\fP (bool bass)"
.br
.RI "enable bass boost option "
.ti -1c
.RI "void \fBmute\fP (bool mute)"
.br
.RI "Toggle mute option, chip will still respond\&. "
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 50 of file tuner\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Tuner::Tuner (uint8_t address, uint8_t SDA, uint8_t SCL, int mcuClock)\fC [inline]\fP"

.PP
Construct a new \fBTuner\fP object\&. 
.PP
\fBParameters\fP
.RS 4
\fIaddress\fP I2C Address 
.br
\fISDA\fP Serial Data Pin 
.br
\fISCL\fP Serial Clock Pin 
.br
\fImcuClock\fP Speed of mcu in HZ 
.RE
.PP

.PP
Definition at line 67 of file tuner\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "void Tuner::bassBoost (bool bass)"

.PP
enable bass boost option 
.PP
\fBParameters\fP
.RS 4
\fIbass\fP 
.RE
.PP

.PP
Definition at line 182 of file tuner\&.cpp\&.
.SS "void Tuner::disableChip ()"

.PP
Disables receiver audio amplifier of chip\&. 
.PP
Definition at line 204 of file tuner\&.cpp\&.
.SS "void Tuner::dumpBuffers (uint8_t read[12], uint8_t write[12]) const"

.PP
dumps buffers from MCU into arrays sent to function 
.PP
\fBParameters\fP
.RS 4
\fIread[12]\fP 12 byte buffer for read data 
.br
\fIwrite[12]\fP 12 byte buffer for write data 
.RE
.PP

.PP
Definition at line 20 of file tuner\&.cpp\&.
.SS "void Tuner::enableChip ()"

.PP
Enables receiver audio amplifier of chip\&. 
.PP
Definition at line 199 of file tuner\&.cpp\&.
.SS "void Tuner::forceMono (bool mono)"

.PP
Force mono signal only\&. 
.PP
\fBParameters\fP
.RS 4
\fImono\fP 
.RE
.PP

.PP
Definition at line 174 of file tuner\&.cpp\&.
.SS "uint8_t Tuner::getSignalStrength ()"

.PP
Get the Signal Strength as a 7-bit int\&. 
.PP
\fBReturns\fP
.RS 4
uint8_t strength 
.RE
.PP

.PP
Definition at line 116 of file tuner\&.cpp\&.
.SS "int Tuner::getStation ()"

.PP
Get channel chip is tuned to currently\&. 
.PP
\fBReturns\fP
.RS 4
int station (Station in MHz * 10) 
.RE
.PP

.PP
Definition at line 82 of file tuner\&.cpp\&.
.SS "uint8_t Tuner::getVolume ()"

.PP
Get the Volume value from buffer\&. 
.PP
\fBReturns\fP
.RS 4
uint8_t volume 
.RE
.PP

.PP
Definition at line 112 of file tuner\&.cpp\&.
.SS "void Tuner::init (void)"

.PP
Resets chip, enables chip, and updates both buffers\&. 
.PP
Definition at line 8 of file tuner\&.cpp\&.
.SS "void Tuner::mute (bool mute)"

.PP
Toggle mute option, chip will still respond\&. 
.PP
\fBParameters\fP
.RS 4
\fImute\fP 
.RE
.PP

.PP
Definition at line 209 of file tuner\&.cpp\&.
.SS "void Tuner::resetChip (bool resetReg)"

.PP
Calling this function preforms soft reset\&. 
.PP
\fBParameters\fP
.RS 4
\fIresetReg\fP Also reset internal buffers as well? 
.RE
.PP

.PP
Definition at line 190 of file tuner\&.cpp\&.
.SS "bool Tuner::seekComplete (void)"

.PP
Status of if scan is complete\&. 
.PP
\fBReturns\fP
.RS 4
true: Scan Complete 
.PP
false: Scan in progress 
.RE
.PP

.PP
Definition at line 146 of file tuner\&.cpp\&.
.SS "void Tuner::seekDown ()"

.PP
Command chip to preform Auto Seek Downwards\&. 
.PP
Definition at line 127 of file tuner\&.cpp\&.
.SS "bool Tuner::seekStatus (void)"

.PP
Status of most recent seek\&. 
.PP
\fBReturns\fP
.RS 4
true: Last seek Succeeded 
.PP
false: Last seek Failed 
.RE
.PP

.PP
Definition at line 137 of file tuner\&.cpp\&.
.SS "void Tuner::seekUp ()"

.PP
Command chip to preform Auto Seek Upwards\&. 
.PP
Definition at line 120 of file tuner\&.cpp\&.
.SS "void Tuner::setSeekSensitiviy (uint8_t threshold)"

.PP
Set the Seek Sensitivity\&. Send a 4-bit number to set the signal to noise ratio of detected signal
.PP
\fBParameters\fP
.RS 4
\fIthreshold\fP (default: 1000) 
.RE
.PP

.PP
Definition at line 152 of file tuner\&.cpp\&.
.SS "void Tuner::setStation (int station)"

.PP
Set the station to tune to\&. 
.PP
\fBParameters\fP
.RS 4
\fIstation\fP (Station in MHz * 10) 
.RE
.PP

.PP
Definition at line 61 of file tuner\&.cpp\&.
.SS "int Tuner::setVolume (uint8_t volume)"

.PP
Set volume of chip audio amplifier\&. Will return error if integer given is out of range
.PP
\fBParameters\fP
.RS 4
\fIvolume\fP (0 to 15) 
.RE
.PP
\fBReturns\fP
.RS 4
int 
.RE
.PP

.PP
Definition at line 99 of file tuner\&.cpp\&.
.SS "bool Tuner::STstatus ()"

.PP
Is the chip receiving a stereo signal? 
.PP
\fBReturns\fP
.RS 4
true: Currently receiving stereo signal 
.PP
false: Currently receiving mono signal 
.RE
.PP

.PP
Definition at line 160 of file tuner\&.cpp\&.
.SS "bool Tuner::TunedStatus ()"

.PP
Is current frequency a station? 
.PP
\fBReturns\fP
.RS 4
true: Chip is Tuned to Station 
.PP
false: Chip is Tuned to Station 
.RE
.PP

.PP
Definition at line 166 of file tuner\&.cpp\&.
.SS "void Tuner::updateDevice (void)"

.PP
Sends updated buffers to tuner chip\&. All chages made with other API calls will not take affect until this command is called\&. 
.PP
Definition at line 15 of file tuner\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for RDA5807M Tuner API from the source code\&.


source/tuner.cpp
#include <stdio.h>
#include "LPC802.h"
#include <math.h>
#include <string.h>
#include "i2c.h"
#include "tuner.h"

void Tuner::init(void) {
	resetChip(true);
	writeDevice();
	readDevice();
	enableChip();
}

void Tuner::updateDevice(void) {
	writeDevice();
	readDevice();
}

void Tuner::dumpBuffers(uint8_t read[12], uint8_t write[12]) const{
	for(int i = 0; i < 12; i++){
		read[i] = TunerReadBuf[i];
		write[i] = TunerWriteBuf[i];
	}
}

inline uint16_t Tuner::StationToBytes(int station) {
	//Example 107.3MHz would be inputted as station = 1703
	//1073 = CHAN_SPACING * CHAN + START FREQ
	uint16_t CHAN = (uint16_t) ((station - START_FREQ) * CHAN_SPACING / 100000);

	return CHAN;
}

inline int Tuner::StationFromBytes(uint16_t CHAN) {
	int station = (int) ((CHAN_SPACING / 100000 * CHAN) + START_FREQ);
	return station;
}

void Tuner::resetBuffer(void) {
	//INITAL VALUES FOR TUNER REGISTER
	TunerWriteBuf[0] = 0b11000000;	//REG 0x02 High Bit
	TunerWriteBuf[1] = 0b00000101;	//REG 0x02 Low Bit

	TunerWriteBuf[2] = 0b00011110;	//REG 0x03 High Bit
	TunerWriteBuf[3] = 0b01000000;	//REG 0x03 Low Bit		DEFAULT CHAN @ BOOT: 99.1MHz ASSUMING US REGION

	TunerWriteBuf[4] = 0b00000000;	//REG 0x04 High Bit
	TunerWriteBuf[5] = 0b00000000;	//REG 0x04 Low Bit

	TunerWriteBuf[6] = 0b11001000;	//REG 0x05 High Bit     1000 seek threshold = 0110
	TunerWriteBuf[7] = 0b00001111;	//REG 0x05 Low Bit

	TunerWriteBuf[8] = 0b00000000;	//REG 0x06 High Bit
	TunerWriteBuf[9] = 0b00000000;	//REG 0x06 Low Bit

	TunerWriteBuf[10] = 0b01000010;	//REG 0x07 High Bit
	TunerWriteBuf[11] = 0b00000010;	//REG 0x07 Low Bit
}

void Tuner::setStation(int station) {
	uint16_t CHAN = StationToBytes(station);

	//MASK AND SET CHANNEL BITS TO 0
	TunerWriteBuf[2] &= ~(WRITECHAN_MASK_HIGH);
	TunerWriteBuf[3] &= ~(WRITECHAN_MASK_LOW);

	//Insert Intended Channel Code	
	//High Byte		Low Byte
	//11111111 		11xxxxxx
	TunerWriteBuf[2] |= (uint8_t) (CHAN >> 2); //11111111  Set upper 8 bits of Chan Select
	TunerWriteBuf[3] |= (uint8_t) (CHAN << 6); //11xxxxxx  Set lower 2 bits of Chan Select

	TunerWriteBuf[3] |= (1UL << 4);	//Set tune bit high to actually tune to selected channel
	writeDevice();
	TunerWriteBuf[3] &= ~(1UL << 4);	//Resets tune bit
										//Tuner reset bit internally
										//but buffer needs to be reset as well
	return;
}

int Tuner::getStation(void) {
	readDevice();
	uint8_t highbyte = TunerReadBuf[0];
	highbyte &= READCHAN_MASK_HIGH;

	uint8_t lowbyte = TunerReadBuf[1];
	lowbyte &= READCHAN_MASK_LOW;

	uint16_t readCHAN = 0;
	readCHAN = lowbyte;
	readCHAN |= (highbyte << 8);

	int station = StationFromBytes(readCHAN);
	return station;
}

//16 Volume Levels, 0 is mute, 15 is full volume
int Tuner::setVolume(uint8_t volume) {
	if (volume >= 0 && volume < 16) {
		//Given that the top half of this bit is reserved
		//we can just equal instead of preform bitwise operations
		TunerWriteBuf[7] &= ~(VOLUME_MASK);
		TunerWriteBuf[7] |= (volume & VOLUME_MASK);
		writeDevice();
		return 0;
	} else {
		return 1;
	}
}

uint8_t Tuner::getVolume() {
	return (TunerWriteBuf[7] & VOLUME_MASK);
}

uint8_t Tuner::getSignalStrength() {
	return (TunerReadBuf[2] >> 1);
}

void Tuner::seekUp() {
	TunerWriteBuf[0] |= SEEKDIR_MASK;	//1 = seekUp
	TunerWriteBuf[0] |= SEEK_TRIGGER_MASK;
	writeDevice();
	TunerWriteBuf[0] &= ~(SEEK_TRIGGER_MASK);
}

void Tuner::seekDown() {
	TunerWriteBuf[0] &= ~(SEEKDIR_MASK);	//0 = seekDown
	TunerWriteBuf[0] |= SEEK_TRIGGER_MASK;
	writeDevice();
	TunerWriteBuf[0] &= ~(SEEK_TRIGGER_MASK);
}

//Did the seek fail?
// 0: Success
// 1: Fail
bool Tuner::seekStatus(void) {
	//readDevice();
	bool status = ((TunerReadBuf[0] & SEEK_FAIL_MASK) >> 5);
	return status;
}

//Did the seek finish / scan through everything
// 0: incomplete
// 1: complete
bool Tuner::seekComplete(void) {
	//readDevice();
	bool status = ((TunerReadBuf[0] & SEEK_COMPLETE_MASK) >> 6);
	return status;
}

void Tuner::setSeekSensitiviy(uint8_t threshold) {
	threshold &= SEEK_THRESHOLD_MASK;		//Ensure threshold is within range
	TunerReadBuf[6] |= SEEK_THRESHOLD_MASK;		//Set bits to xxxx1111
	TunerReadBuf[6] &= ~(SEEK_THRESHOLD_MASK); 	//Clear bits to xxxx0000
	TunerReadBuf[6] |= threshold;				//Assign to bits
	writeDevice();								//Update Device
}

bool Tuner::STstatus() {
	bool status = false;
	status = ((TunerReadBuf[0] & STERO_INDICATOR_MASK) >> 2);
	return status;
}

bool Tuner::TunedStatus() {
	readDevice();
	if ((TunerReadBuf[2] & FM_TRUE_MASK) > 0)
		return true;
	else
		return false;
}

void Tuner::forceMono(bool mono) {
	if (mono)
		TunerWriteBuf[0] |= (MONO_MASK);
	else
		TunerWriteBuf[0] &= ~(MONO_MASK);
	writeDevice();
}

void Tuner::bassBoost(bool bass) {
	if (bass)
		TunerWriteBuf[0] |= (BASS_MASK);
	else
		TunerWriteBuf[0] &= ~(BASS_MASK);
	writeDevice();
}

void Tuner::resetChip(bool resetReg) {
	TunerWriteBuf[1] |= RESET_MASK;
	writeDevice();
	delayMilli(300);
	TunerWriteBuf[1] &= ~(RESET_MASK);
	if (resetReg)
		resetBuffer();
}

void Tuner::enableChip() {
	TunerWriteBuf[1] |= ENABLE_MASK;
	writeDevice();
}

void Tuner::disableChip() {
	TunerWriteBuf[1] &= ~(ENABLE_MASK);
	writeDevice();
}

void Tuner::mute(bool mute) {
	if (mute)
		TunerWriteBuf[0] |= (MUTE_MASK);
	else
		TunerWriteBuf[0] &= ~(MUTE_MASK);
	writeDevice();

}

void Tuner::delayMicro(int value) {
	int maxVal = ((value / 2) * _mcuClock / 12000000);
	for (int i = 0; i < maxVal; i++)
		asm("nop");
}

void Tuner::delayMilli(int value) {
	int maxVal = ((value / 2) * _mcuClock / 11600);
	for (int i = 0; i < maxVal; i++)
		asm("nop");
}

void Tuner::readDevice(void) {
#ifndef DEBUG
	I2CChangePin(_SDApin, _SCLpin);
	/*
	 WaitI2CMasterState(I2C0, I2C_STAT_MSTST_IDLE); // Wait for the master state to be idle
	 I2C0->MSTDAT = (TUNER_I2CAddress<<1) | 0; // Address with 0 for RWn bit (WRITE)
	 I2C0->MSTCTL = MSTCTL_START; //Start the transaction by setting the MSTSTART bit to 1 in the Master control register.

	 WaitI2CMasterState(I2C0, I2C_STAT_MSTST_TXRDY); // Wait for the address to be ACK'd
	 I2C0->MSTDAT = 0x00;							//Put LM75 into config mode (I THINK)
	 I2C0->MSTCTL = MSTCTL_CONTINUE;
	 */
	WaitI2CMasterState(I2C0, I2C_STAT_MSTST_IDLE);
	I2C0->MSTDAT = (_addr << 1) | 1; // Address with 1 for RWn bit (READ)
	I2C0->MSTCTL = MSTCTL_START; 		// Start the transaction by setting the
										// MSTSTART bit to 1 in the Master control register.
	for (int i = 0; i < TOTALREAD_REG; i++) { //6 registers to read from
		//WaitI2CMasterState(I2C0, I2C_STAT_MSTST_TXRDY); // Wait for the address to be ACK'd
		//WaitI2CMasterState(I2C0, I2C_STAT_MSTST_IDLE);
		//WaitI2CMasterState(I2C0, I2C_STAT_MSTST_RXRDY);
		while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_RXRDY)
			; // WAIT FOR THIS CONDITION TO BE TRUE // MASTER_STATE_MASK is (0x7<<1)
		TunerReadBuf[2 * i] = I2C0->MSTDAT;	//STORE HIGH BIT
		I2C0->MSTCTL = MSTCTL_CONTINUE;

		WaitI2CMasterState(I2C0, I2C_STAT_MSTST_RXRDY); // Wait for the address to be ACK'd
		// MASTER_STATE_MASK is (0x7<<1)
		while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_RXRDY)
			; //WAIT FOR THIS CONDITION TO BE TRUE
		TunerReadBuf[(2 * i) + 1] = I2C0->MSTDAT; //STORE LOW BIT

		if (i != TOTALREAD_REG - 1)
			I2C0->MSTCTL = MSTCTL_CONTINUE;
	}
	I2C0->MSTCTL = MSTCTL_STOP;

	// MASTER_STATE_MASK is (0x7<<1)
	while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_IDLE)
		;
#endif
}

//Uses buffer and pushes it to tuner
void Tuner::writeDevice(void) {
#ifndef DEBUG
	I2CChangePin(_SDApin, _SCLpin);
	WaitI2CMasterState(I2C0, I2C_STAT_MSTST_IDLE); // Wait for the master state to be idle
	I2C0->MSTDAT = (_addr << 1) | 0; // Address with 0 for RWn bit (WRITE)
	I2C0->MSTCTL = MSTCTL_START; //Start the transaction by setting the MSTSTART bit to 1 in the Master control register.
	for (int i = 0; i < TOTALWRITE_REG; i++) {
		WaitI2CMasterState(I2C0, I2C_STAT_MSTST_TXRDY); // Wait for the address to be ACK'd
		while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_TXRDY)
			; //WAIT FOR THIS CONDITION TO BE TRUE
		I2C0->MSTDAT = TunerWriteBuf[2 * i]; //Put LM75 into config mode (I THINK)
		I2C0->MSTCTL = MSTCTL_CONTINUE;

		WaitI2CMasterState(I2C0, I2C_STAT_MSTST_TXRDY); // Wait for the High Bit to be ACK'd
		while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_TXRDY)
			; //WAIT FOR THIS CONDITION TO BE TRUE
		I2C0->MSTDAT = TunerWriteBuf[(2 * i) + 1]; //Put LM75 into config mode (I THINK)

		//if(i != TOTALWRITE_REG-1)
		I2C0->MSTCTL = MSTCTL_CONTINUE;

	}
	WaitI2CMasterState(I2C0, I2C_STAT_MSTST_TXRDY); // Wait for the Low bit to be ACK'd
	//while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_TXRDY); //WAIT FOR THIS CONDITION TO BE TRUE
	I2C0->MSTCTL = MSTCTL_STOP;
	while ((I2C0->STAT & MASTER_STATE_MASK) != I2C_STAT_MSTST_IDLE)
		;
#endif
}
